cuML/src/kmeans/kmeans.cuh

```diff
@@ -0,0 +1,216 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <linalg/matrix_vector_op.h>
+#include <linalg/reduce_rows_by_key.h>
+#include <linalg/mean_squared_error.h>
+#include <linalg/binary_op.h>
+#include <distance/distance.h>
+#include <matrix/gather.h>
+#include <random>
+#include <random/rng.h>
+#include <random/permute.h>
+
+#include <numeric>
+#include <thrust/fill.h>
+#include <thrust/for_each.h>
+#include <thrust/scan.h>
+#include <thrust/execution_policy.h>
+
+#include <ml_cuda_utils.h>
+#include <common/tensor.hpp>
+#include <common/cumlHandle.hpp>
+#include <common/device_buffer.hpp>
+#include <common/host_buffer.hpp>
+#include <cuML.hpp>
+
+#include "kmeans.hpp"
+
+namespace ML {
+using namespace MLCommon;
+
+/**
+ * @tparam DataT 
+ * @tparam IndexT
+ */
+template <typename DataT,
+	  typename IndexT = int>
+class KMeans{
+public:
+
+    struct Options{
+	int seed;
+	int oversampling_factor;
+	Random::GeneratorType gtype;
+	bool inertia_check;
+    public:
+	Options(){
+	    seed = -1;
+	    oversampling_factor = 0;
+	    gtype = Random::GeneratorType::GenPhilox;
+	    inertia_check = false;
+	}
+    };
+
+    
+    /**
+     * @param[in] cumlHandle  The handle to the cuML library context that manages the CUDA resources.
+     * @param[in] n_clusters  The number of clusters to form as well as the number of centroids to generate (default:8).
+     * @param[in] metric      Metric to use for distance computation. Any metric from MLCommon::Distance::DistanceType can be used
+     * @param[in] init        Method for initialization, defaults to k-means++:
+     *                            - InitMethod::KMeansPlusPlus (k-means++): Use scalable k-means++ algorithm to select the initial cluster centers.
+     *                            - InitMethod::Random (random): Choose 'n_clusters' observations (rows) at random from the input data for the initial centroids.
+     *                            - InitMethod::Array (ndarray): Use the user provided input for initial centroids, it should be set using setCentroids method. 
+     * @param[in] max_iter    Maximum number of iterations for the k-means algorithm.
+     * @param[in] tol         Relative tolerance with regards to inertia to declare convergence.
+     * @param[in] seed        Seed to the random number generator.
+     */
+    KMeans(const ML::cumlHandle_impl &cumlHandle,
+	   int n_clusters = 8,
+	   int metric = static_cast<MLCommon::Distance::DistanceType>(0),
```

Comment:
As the type definition of `MLCommon::Distance::DistanceType` is available here why can't we use `MLCommon::Distance::DistanceType` directly instead of `int`?

Reply:
We can use enum directly here and I will make the necessary change. But JFYI, at kmeans.hpp we can't use it because cython currently is passing `int` and once we expose `DistanceType` to cython then we can pass `DistanceType` uniformly across all layer.